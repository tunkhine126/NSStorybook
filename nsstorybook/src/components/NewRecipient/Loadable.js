/**
 *
 * Asynchronously loads the component for New Recipients
 *
 */
import React from 'react';
import loadable from 'utils/loadable';
import Loading from 'components/shared/Loading';

export default loadable(() => import('./index'), {
  fallback: <Loading />,
});
