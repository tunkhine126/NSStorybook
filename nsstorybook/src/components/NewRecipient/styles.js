import { makeStyles } from '@material-ui/core/styles';
import { green, amber } from '@material-ui/core/colors';
import colors from '../../global-styles';
import styled from 'styled-components';

const separatorGray = '#BBBBBB';

export const commonStyles = makeStyles(theme => ({
  darkText: {
    color: colors.TEXT.dark,
    fontSize: '14px',
    fontWeight: 'normal',
  },
  generalContainer: {
    minHeight: '310px',
    borderBottom: `1px solid ${separatorGray}`,
  },
  header: {
    fontSize: '28px',
  },
  input: {
    width: '100%',
    maxWidth: '400px',
  },
  fieldsContainer: {
    padding: '25px 0',
    flexGrow: 0,
  },
  root: {
    flexGrow: 1,
    display: 'flex',
    marginTop: '1.5rem',
    borderTop: `1px solid ${separatorGray}`,
  },
  sectionHeader: {
    fontSize: '16px',
    fontWeight: 'bold',
    color: colors.INTERFACE.ui7,
  },
  spacing: {
    marginBottom: '5px',
  },
  selectSpacing: {
    marginBottom: '27px',
  },
  success: {
    backgroundColor: green[600],
  },
  error: {
    backgroundColor: theme.palette.error.dark,
  },
  info: {
    backgroundColor: theme.palette.primary.main,
  },
  warning: {
    backgroundColor: amber[700],
  },
  linkStyle: {
    textTransform: 'uppercase',
    fontWeight: '700',
    fontSize: '12px',
    letterSpacing: '1px',
  },
  optionalFields: {
    paddingTop: '25px',
  },
  required: {
    minHeight: '202px',
    borderBottom: '1px solid #BBBBBB',
    [theme.breakpoints.down('md')]: {
      minHeight: '225px',
    },
  },
}));

export const InputWrapper = styled.div`
  .MuiInputBase-root {
    height: 58px;
  }
  .MuiFormControl-root {
    max-width: 66%;
  }
  .MuiFilledInput-multiline {
    height: 100px;
  }
`;
