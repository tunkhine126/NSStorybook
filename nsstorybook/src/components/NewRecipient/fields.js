/**
 *
 * Fields
 *
 */

import React, { useState } from 'react';
import PropTypes from 'prop-types';

import Grid from '@material-ui/core/Grid';

import NSFields from '../NSFields';

import { required } from 'utils/validations';
import { commonStyles } from './styles';

export function Fields({
  fields,
  showPreFixId,
  preFixSeparator,
  parentInternalId,
  preSetValue,
  fieldsRequired = false,
}) {
  let requiredProps = {};
  const classes = commonStyles();

  if (fieldsRequired) {
    requiredProps = {
      validateOnChange: true,
      validateOnBlur: true,
      required: true,
      validate: required,
    };
  }

  const [fieldState, setFieldState] = useState(
    fields.reduce((o, key) => ({ ...o, [key.name]: { item: [] } }), {})
  );

  const handleFieldChange = (field, value) =>
    setFieldState(state => ({ ...state, [field]: { item: value } }));

  const Prefix = () => (
    <p style={{ width: 'max-content' }}>{parentInternalId + preFixSeparator}</p>
  );

  const fieldsToRender =
    fields.length > 0 &&
    fields.map((f, idx) => {
      const selectFields = ['single_select', 'multi_select', 'likert'];
      if (selectFields.includes(f.field_type)) {
        return (
          <NSFields
            key={idx}
            fieldType={f.field_type}
            field={`${f.name}_custom`}
            label={f.contexts.us_en.label}
            values={fieldState[f.name]}
            options={f.option_values.map(o => ({ value: o }))}
            onchange={v => handleFieldChange(f.name, v)}
            {...requiredProps}
          />
        );
      }
      return (
        <NSFields
          key={idx}
          fieldType={f.field_type}
          field={`${f.name}_custom`}
          label={f.contexts.us_en.label}
          {...requiredProps}
        />
      );
    });
  return (
    <Grid item className={classes.spacing}>
      {fieldsRequired && (
        <NSFields
          fieldType="short_text"
          field="name"
          label="Name"
          {...requiredProps}
        />
      )}

      {!fieldsRequired && (
        <>
          <NSFields
            fieldType="photo"
            field="base64Avatar"
            label="Avatar"
            {...requiredProps}
          />

          <NSFields
            fieldType="short_text"
            field="internalId"
            label="Internal Id"
            position="start"
            initialValue={preSetValue}
            showPrefix={showPreFixId}
            icon={Prefix}
            {...requiredProps}
          />

          <NSFields
            fieldType="paragraph_text"
            field="description"
            label="Description"
            {...requiredProps}
          />
        </>
      )}

      {fieldsToRender}
    </Grid>
  );
}

Fields.propTypes = {
  fields: PropTypes.array.isRequired,
  fieldsRequired: PropTypes.bool,
  showPreFixId: PropTypes.bool,
  preFixSeparator: PropTypes.string,
  parentInternalId: PropTypes.string,
  preSetValue: PropTypes.string,
};

export default Fields;
