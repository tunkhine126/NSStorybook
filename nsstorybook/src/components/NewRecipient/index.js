/**
 *
 * NewRecipient
 *
 */

import React, { useState, useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import { useQuery, useMutation } from '@apollo/react-hooks';
import { injectIntl, intlShape } from 'react-intl';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { Form } from 'informed';
import queryString from 'query-string';

import Grid from '@material-ui/core/Grid';

// Queries
import {
  GET_RECIEPIENT_DEFINITION,
  GET_RECIEPIENT_INTERNAL_ID,
} from 'graphql/queries/recipients';

// Mutations
import { CREATE_RECIPIENT } from 'graphql/mutations/families/families-mutation';

// Components
import { showToaster as toast } from 'containers/App/actions';
import { generalErrorHandler } from 'utils/error-handler';
import Content from 'components/shared/Content';
import ActionPanel from 'components/shared/ActionPanel';
import Loading from 'components/shared/Loading';
import ViewHeader from 'components/shared/ViewHeader';
import Notification from 'components/shared/Notification';
import globalMessages from 'messages';
import { filterObj } from '../../utils/helpers';
import { commonStyles, InputWrapper } from './styles';
import Fields from './fields';

function NewRecipient({ intl, location, showToaster, ...rest }) {
  const formApiRef = useRef();
  const classes = commonStyles();
  const [fields, setFields] = useState({ required: [], optional: [] });
  const [loadingState, setloadingState] = useState(false);
  const [redirectState, setRedirectState] = useState(false);
  const [messageState, setmessageState] = useState({
    message: '',
    open: false,
    type: 'success',
  });

  const messageHandler = (event, reason) => {
    if (reason === 'clickaway') return;
    setmessageState({ open: false });
  };

  const handleFormReset = () => {
    formApiRef.current.reset();
  };

  const { loading, error, data } = useQuery(GET_RECIEPIENT_DEFINITION, {
    variables: {
      uuid: location.search.includes('&rdId=')
        ? queryString.parse(location.search).rdId
        : null,
    },
  });

  const {
    loading: internalIdLoading,
    error: internalIdError,
    data: internalIdData,
  } = useQuery(GET_RECIEPIENT_INTERNAL_ID, {
    variables: {
      uuid: location.search.includes('pId=')
        ? queryString.parse(location.search).pId
        : null,
    },
  });

  const [createRecipient] = useMutation(CREATE_RECIPIENT);

  useEffect(() => {
    const rd = data && data.recipientDefinition;
    const id = internalIdData && internalIdData.recipient;
    if (rd && id) {
      const customFields = rd.customFields ? [...rd.customFields] : [];

      setFields({
        showPreFixId: rd.prefixInternalIds,
        preFixSeparator: rd.prefixSeparator,
        preSetValue: id.children && (id.children.length + 1).toString(),
        parentInternalId: id.internalId,
        required: customFields.filter(
          f => f.on_create === 'require' || f.on_create === 'required'
        ),
        optional: customFields.filter(f => f.on_create === 'show'),
      });
    }
  }, [data, internalIdData]);

  const handleSubmit = async formData => {
    const custom = '_custom';
    const defaultFields = filterObj(formData, d => !d.includes(custom));
    let customFields = filterObj(formData, d => d.includes(custom));
    customFields = Object.keys(customFields)
      .map(key => key.replace(custom, ''))
      .reduce(
        (res, key) =>
          Object.assign(res, { [key]: customFields[`${key}${custom}`] }),
        {}
      );

    try {
      setloadingState(true);
      const response = await createRecipient({
        variables: {
          ...defaultFields,
          customFields: JSON.stringify(customFields),
          parentUuid: queryString.parse(location.search).pId,
          recipientDefinitionUuid: queryString.parse(location.search).rdId,
        },
      });
      if (response.data) {
        setRedirectState(true);
      }
    } catch (error) {
      const { graphQLErrors, networkError } = error;
      setloadingState(false);

      if (graphQLErrors.length > 0) {
        const [{ message }] = graphQLErrors;

        setmessageState({
          message,
          open: true,
          type: 'error',
        });
      }

      if (networkError) {
        const { message } = networkError;
        const failedConnection = message.includes('Failed to fetch');
        if (failedConnection) {
          setmessageState({
            message: intl.formatMessage(globalMessages.networkError),
            open: true,
            type: 'error',
          });
        }
      }
    }
  };

  if (error || internalIdError) {
    return generalErrorHandler(error || internalIdError, {
      ...rest,
    });
  }

  if (loading || internalIdLoading) {
    return <Loading />;
  }

  // Redirect user if query params are missing
  if (
    !location.search.includes('pId=') ||
    !location.search.includes('&rdId=') ||
    !location.search.includes('&name=') ||
    !location.search.includes('&path=')
  ) {
    return rest.history.go(-1);
  }

  if (redirectState) {
    showToaster();

    return (
      <Redirect
        to={{
          pathname: queryString.parse(location.search).path,
        }}
      />
    );
  }

  return (
    <>
      <ViewHeader
        component={queryString.parse(location.search).name}
        body={queryString.parse(location.search).name}
      />

      <div className={classes.root}>
        <Grid container>
          <Notification
            status={messageState.open}
            message={messageState.message}
            classes={classes}
            messageHandler={messageHandler}
            duration={3000}
            position={{ vertical: 'top', horizontal: 'right' }}
            type={messageState.type}
          />
          <Grid item xs={9} xl={10} className={classes.fieldsContainer}>
            <InputWrapper>
              <Form
                getApi={formApi => {
                  formApiRef.current = formApi;
                }}
                id="add-new-recipient"
                onSubmit={formData => handleSubmit(formData)}
              >
                {fields.required.length > 0 && (
                  <Content
                    subHeader={`1 ${intl.formatMessage(
                      globalMessages.required
                    )}`}
                    body={<Fields fields={fields.required} fieldsRequired />}
                    containerClass={classes.required}
                    headerClass={classes.sectionHeader}
                  />
                )}

                {fields.optional.length > 0 && (
                  <Content
                    subHeader={`2 ${intl.formatMessage(
                      globalMessages.optionalFields
                    )}`}
                    body={
                      <Fields
                        fields={fields.optional}
                        showPreFixId={fields.showPreFixId}
                        preFixSeparator={fields.preFixSeparator}
                        parentInternalId={fields.parentInternalId}
                        preSetValue={fields.preSetValue}
                      />
                    }
                    containerClass={classes.optionalFields}
                    headerClass={classes.sectionHeader}
                  />
                )}
              </Form>
            </InputWrapper>
          </Grid>
          <ActionPanel
            action="addRecipientChild"
            type="submit"
            formId="add-new-recipient"
            resetForm={handleFormReset}
            loading={loadingState}
            disabled={loadingState}
          />
        </Grid>
      </div>
    </>
  );
}

NewRecipient.propTypes = {
  intl: intlShape.isRequired,
  location: PropTypes.object.isRequired,
  showToaster: PropTypes.func,
};

const intlNewRecipient = injectIntl(NewRecipient);

export function mapDispatchToProps(dispatch) {
  return {
    showToaster: () => dispatch(toast()),
    dispatch,
  };
}

export default connect(
  null,
  mapDispatchToProps
)(intlNewRecipient);
