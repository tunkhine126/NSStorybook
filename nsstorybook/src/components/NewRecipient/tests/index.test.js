/**
 *
 * Tests for NewRecipient
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render } from '@testing-library/react';
// import 'jest-dom/extend-expect'; // add some helpful assertions

import TestingComponent from 'components/shared/Testing';
import NewRecipient from '../Loadable';

describe('<NewRecipient />', () => {
  const location = {
    search:
      'pId=07c538ec-f5c8-11e9-a1ad-22000be1c01d&rdId=60618bd0-5641-11e9-9966-22000bd4493b',
  };

  const invalidLocation = {
    search:
      'pI=07c538ec-f5c8-11e9-a1ad-22000be1c01d&dId=60618bd0-5641-11e9-9966-22000bd4493b',
  };

  it('Expect no errors rending Fields component', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <TestingComponent>
        <NewRecipient location={location} />
      </TestingComponent>
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should redirect user to family list', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <TestingComponent>
        <NewRecipient location={invalidLocation} />
      </TestingComponent>
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <TestingComponent>
        <NewRecipient location={location} />
      </TestingComponent>
    );
    expect(firstChild).toMatchSnapshot();
  });
});
