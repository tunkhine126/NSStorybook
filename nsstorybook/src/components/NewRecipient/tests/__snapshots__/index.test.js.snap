// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewRecipient /> Should render and match the snapshot 1`] = `
<div
  class="MuiGrid-root makeStyles-container-257 MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
  >
    <div
      aria-label="animation"
      role="button"
      style="width: 50px; height: 50px; overflow: hidden; margin: 0px auto; outline: none;"
      tabindex="0"
      title=""
    >
      <svg
        height="500"
        preserveAspectRatio="xMidYMid slice"
        style="width: 100%; height: 100%; transform: translate3d(0,0,0);"
        viewBox="0 0 500 500"
        width="500"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <clippath
            id="__lottie_element_62"
          >
            <rect
              height="500"
              width="500"
              x="0"
              y="0"
            />
          </clippath>
        </defs>
        <g
          clip-path="url(#__lottie_element_62)"
        >
          <g
            style="display: none;"
          >
            <g>
              <path />
            </g>
          </g>
          <g
            style="display: none;"
          >
            <g>
              <path />
            </g>
          </g>
          <g
            style="display: none;"
          >
            <g>
              <path />
            </g>
          </g>
          <g
            opacity="1"
            style="display: block;"
            transform="matrix(1,0,0,1,249,250)"
          >
            <g
              opacity="1"
              transform="matrix(1,0,0,1,0,0)"
            >
              <path
                d=" M-78,-233.5 C-78,-233.5 -218,-233.5 -218,-233.5 C-218,-233.5 100.25,232 100.25,232 C100.25,232 217.10000610351562,231.89599609375 217.10000610351562,231.89599609375 C217.10000610351562,231.89599609375 217,200.5 217,200.5 C217,200.5 -78,-233.5 -78,-233.5z"
                fill="rgb(0,0,0)"
                fill-opacity="1"
              />
            </g>
          </g>
          <g
            opacity="1"
            style="display: block;"
            transform="matrix(1.0184400081634521,0,0,1.1380000114440918,575.5169677734375,220.61326599121094)"
          >
            <g
              opacity="1"
              transform="matrix(1,0,0,1,-161.76699829101562,25.211000442504883)"
            >
              <path
                d=" M54.23350143432617,-204.00999450683594 C54.23350143432617,-204.00999450683594 54.23350143432617,204.00999450683594 54.23350143432617,204.00999450683594 C54.23350143432617,204.00999450683594 -54.23350143432617,204.00999450683594 -54.23350143432617,204.00999450683594 C-54.23350143432617,204.00999450683594 -54.23350143432617,-204.00999450683594 -54.23350143432617,-204.00999450683594 C-54.23350143432617,-204.00999450683594 54.23350143432617,-204.00999450683594 54.23350143432617,-204.00999450683594z"
                fill="rgb(0,0,0)"
                fill-opacity="1"
              />
            </g>
          </g>
          <g
            opacity="1"
            style="display: block;"
            transform="matrix(-1,0,0,-1.0229300260543823,-73.03399658203125,493.742431640625)"
          >
            <g
              opacity="1"
              transform="matrix(1,0,0,1,-161.76699829101562,120.48999786376953)"
            >
              <path
                d=" M54.23350143432617,-109.01000213623047 C54.23350143432617,-109.01000213623047 54.23350143432617,109.01000213623047 54.23350143432617,109.01000213623047 C54.23350143432617,109.01000213623047 -54.23350143432617,109.01000213623047 -54.23350143432617,109.01000213623047 C-54.23350143432617,109.01000213623047 -54.23350143432617,-109.01000213623047 -54.23350143432617,-109.01000213623047 C-54.23350143432617,-109.01000213623047 54.23350143432617,-109.01000213623047 54.23350143432617,-109.01000213623047z"
                fill="rgb(0,0,0)"
                fill-opacity="1"
              />
            </g>
          </g>
        </g>
      </svg>
    </div>
  </div>
</div>
`;
